crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -337,163 +338,463 @@ pub fn decompose_dense(
-        let mut z = Array2::<bool>::uninit([paulis.len(), num_qubits]);
-        let mut x = Array2::<bool>::uninit([paulis.len(), num_qubits]);
-        let mut phases = Array1::<u8>::uninit(paulis.len());
-        for (i, paulis) in paulis.drain(..).enumerate() {
-            let mut phase = 0u8;
-            for (j, pauli) in paulis.into_iter().rev().enumerate() {
-                match pauli {
-                    Pauli::I => {
-                        z[[i, j]].write(false);
-                        x[[i, j]].write(false);
-                    }
-                    Pauli::X => {
-                        z[[i, j]].write(false);
-                        x[[i, j]].write(true);
-                    }
-                    Pauli::Y => {
-                        z[[i, j]].write(true);
-                        x[[i, j]].write(true);
-                        phase = phase.wrapping_add(1);
-                    }
-                    Pauli::Z => {
-                        z[[i, j]].write(true);
-                        x[[i, j]].write(false);
-                    }
+    let (stack, mut out_list, mut scratch) = decompose_first_level(operator.as_array(), num_qubits);
+    decompose_middle_levels(stack, &mut out_list, &mut scratch, num_qubits);
+    let out = decompose_last_level(&mut out_list, &scratch, num_qubits, tolerance)?;
+    Ok(ZXPaulis {
+        z: PyArray1::from_vec_bound(py, out.z)
+            .reshape([out.phases.len(), num_qubits])?
+            .into(),
+        x: PyArray1::from_vec_bound(py, out.x)
+            .reshape([out.phases.len(), num_qubits])?
+            .into(),
+        phases: PyArray1::from_vec_bound(py, out.phases).into(),
+        coeffs: PyArray1::from_vec_bound(py, out.coeffs).into(),
+    })
+}
+
+/// Apply the matrix-addition decomposition at the first level.  This is split out because it acts
+/// on an `ArrayView2`, and is responsible for populating the initial scratch space.  We can't
+/// write over the operator the user gave us (it's not ours to do that to), and anyway, we want to
+/// drop to a chunk of memory that we can 100% guarantee is contiguous, so we can elide all the
+/// stride checking.
+fn decompose_first_level(
+    in_op: ArrayView2<Complex64>,
+    num_qubits: usize,
+) -> (Vec<PauliLocation>, Vec<PauliLocation>, Vec<Complex64>) {
+    let side = 1 << num_qubits;
+    let mut stack = Vec::<PauliLocation>::with_capacity(4);
+    let mut out_list = Vec::<PauliLocation>::new();
+    let mut scratch = Vec::<Complex64>::with_capacity(side * side);
+    match num_qubits {
+        0 => {}
+        1 => {
+            // If we've only got one qubit, we just want to copy the data over in the correct
+            // continuity and let the base case of the iteration take care of outputting it.
+            scratch.extend(in_op.iter());
+            out_list.push(PauliLocation::begin(num_qubits));
+        }
+        _ => {
+            unsafe { scratch.set_len(side * side) };
```

Comment:
Is there a way to test that there is no memory leakage?

Reply:
I can write a Rust-space test, since that'll cause all this code to be run under Miri, which checks this kind of thing.  In the mean time I made this `scratch.set_len(scratch.capacity())` so it's clear that it's not setting it wider than is allocated.